code/go/pkg/validator/limits_test.go

```diff
@@ -0,0 +1,382 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package validator
+
+import (
+	_ "embed"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/package-spec/code/go/internal/spectypes"
+)
+
+func TestLimitsValidation(t *testing.T) {
+	cases := []struct {
+		title string
+		fsys  fs.FS
+		valid bool
+	}{
+		{
+			title: "all good",
+			fsys:  newMockFS().Good(),
+			valid: true,
+		},
+		{
+			title: "configurationSizeLimit exceeded",
+			fsys: newMockFS().Good().Override(func(o *overrideFS) {
+				o.File("manifest.yml").WithSize(10 * spectypes.MegaByte)
+			}),
+			valid: false,
+		},
+		{
+			title: "sizeLimit exceeded",
+			fsys: newMockFS().Good().Override(func(o *overrideFS) {
+				o.File("docs/README.md").WithSize(200 * spectypes.MegaByte)
+			}),
+			valid: false,
+		},
+		{
+			title: "totalSizeLimit exceeded",
+			fsys: newMockFS().Good().WithFiles(
+				newMockFile("docs/other.md").WithSize(140*spectypes.MegaByte),
+				newMockFile("docs/someother.md").WithSize(140*spectypes.MegaByte),
+			),
+			valid: false,
+		},
+		{
+			title: "totalContentsLimit exceeded",
+			fsys: newMockFS().Good().Override(func(o *overrideFS) {
+				o.File("docs").WithGeneratedFiles(70000, ".md", 512*spectypes.Byte)
+			}),
+			valid: false,
+		},
+		{
+			title: "relativePathSizeLimit exceeded",
+			fsys: newMockFS().Good().Override(func(o *overrideFS) {
+				o.File("img/kibana-system.png").WithSize(10 * spectypes.MegaByte)
+			}),
+			valid: false,
+		},
+		{
+			title: "data streams limit exceeded",
+			fsys: newMockFS().Good().Override(func(o *overrideFS) {
+				o.MultiplyFile("data_stream", "foo", 1000)
+			}),
+			valid: false,
+		},
+		{
+			title: "fieldsPerDataStreamLimit exceeded",
+			fsys: newMockFS().Good().WithFiles(
+				newMockFile("data_stream/foo/fields/many-fields.yml").WithContent(generateFields(1500)),
+			),
+			valid: false,
+		},
+		{
+			title: "config template sizeLimit exceeded",
+			fsys: newMockFS().Good().WithFiles(
+				newMockFile("agent/input/stream.yml.hbs").WithSize(6 * spectypes.MegaByte),
+			),
+			valid: false,
+		},
+		{
+			title: "ingest pipeline sizeLimit exceeded",
+			fsys: newMockFS().Good().WithFiles(
+				newMockFile("elasticsearch/ingest_pipeline/pipeline.yml").WithContent("---\n").WithSize(4 * spectypes.MegaByte),
+			),
+			valid: false,
+		},
+		{
+			title: "ignore developer files",
+			fsys: newMockFS().Good().WithFiles(
+				newMockFile("_dev/deploy/docker/entrypoint.sh").WithSize(2048 * spectypes.MegaByte),
+			),
+			valid: true,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.title, func(t *testing.T) {
+			err := ValidateFromFS("test-package", c.fsys)
+			if c.valid {
+				assert.NoError(t, err)
+			} else {
+				assert.Error(t, err)
+			}
+		})
+	}
+}
+
+var _ fs.FS = &mockFS{}
+
+type mockFS struct {
+	root *mockFile
+}
+
+func newMockFS() *mockFS {
```

Comment:
I know there is this [project](https://github.com/spf13/afero), but not sure if it isn't an overkill. WDYT?

Reply:
I considered it, but this provides a full in-memory filesystem. I didn't find it so suited for mocking responses. For example I think that with aferofs I would need to create files of a given size, and not only mock the result of `info.Size()`. Should I give it another try?

Reply:
No, it's a good justification then. It looks you found the niche :) Let's leave it as is.